### PR TITLE
Intercept errors thrown by synchronous Completers in image resolution.

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

If an error is added to a synchronous completer before a listener has been  added, it can throw an error both into the zone and up the stack. Thus, it  looks like the error has been caught, but it is infact also bubbling to the zone. Since we cannot prevent all usage of Completer.sync here, or rather that changing them would be too breaking, we instead hook into the same zone mechanism to intercept the uncaught error and deliver it to the image stream's error handler. Note that these errors will be duplicated, hence the need for the `didError` flag.

## Related Issues

Fixes internal issue b/120841734

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes tests for *all* changed/updated/fixed behaviors (See [Test Coverage]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by jonahwilliams** [comment]
cc @goderbauer, integration tests in progress

**Comment by jonahwilliams** [comment]
Integration test added, we unfortunately cannot run the driver extensions because it changes the behavior of errors.

File: packages/flutter/lib/src/painting/image_provider.dart [hunk 1]
```diff
@@ -281,27 +286,38 @@ abstract class ImageProvider<T> {
...
+    // image stream's error handler. Note that these errors will be duplicated,
+    // hence the need for the `didError` flag.
+    final Zone dangerZone = Zone.current.fork(
+      specification: ZoneSpecification(
+        handleUncaughtError: (Zone zone, ZoneDelegate delegate, Zone parent, Object error, StackTrace stackTrace) {
+          handleError(error, stackTrace);
+        }
+      )
+    );
+    dangerZone.run(() {
```
[the remark this answers is not in the document]
Comment (yjbanov) [reply packages/flutter/lib/src/painting/image_provider.dart]: Also, `run<void>`?

File: dev/integration_tests/image_loading/lib/main.dart [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+      context: 'In the Zone handleUncaughtError handler',
+      silent: false,
+    ));
+  });
+  when(httpClient.getUrl(any)).thenAnswer((Invocation invocation) {
+    final Completer<HttpClientRequest> completer = Completer<HttpClientRequest>.sync();
+    completer.completeError(Error());
+    return completer.future;
+  });
+  Zone.current.fork(specification: specification).run(() {
```
Comment (yjbanov) [inline dev/integration_tests/image_loading/lib/main.dart]: `run<void>`?

File: packages/flutter/test/painting/image_provider_test.dart [hunk 3]
```diff
@@ -88,30 +90,100 @@ void main() {
...
+      FlutterError.onError = (FlutterErrorDetails details) {
+        caughtError.complete(false);
+      };
+      final ImageStream result = imageProvider.resolve(ImageConfiguration.empty);
+      result.addListener((ImageInfo info, bool syncCall) {
+      }, onError: (dynamic error, StackTrace stackTrace) {
+        caughtError.complete(true);
+      });
+      expect(await caughtError.future, true);
+      print('done');
```
Comment (yjbanov) [inline packages/flutter/test/painting/image_provider_test.dart]: Do you need to print in the test?

File: packages/flutter/test/painting/image_provider_test.dart [hunk 4]
```diff
@@ -88,30 +90,100 @@ void main() {
...
+    final Zone testZone = Zone.current.fork(specification: ZoneSpecification(
+      handleUncaughtError: (_, __, ___, ____, _____) {
+        uncaught = true;
+      }
+    ));
+    await testZone.run(() async {
+      final ImageProvider imageProvider = LoadErrorImageProvider();
+      final Completer<bool> caughtError = Completer<bool>();
+      FlutterError.onError = (FlutterErrorDetails details) {
+        caughtError.complete(false);
```
Comment (yjbanov) [inline packages/flutter/test/painting/image_provider_test.dart]: Do you want this to be called at all? Maybe throw?

File: packages/flutter/test/painting/image_provider_test.dart [hunk 5]
```diff
@@ -88,30 +90,98 @@ void main() {
...
+      }, onError: (dynamic error, StackTrace stackTrace) {
+        caughtError.complete(true);
+      });
+      expect(await caughtError.future, true);
     });
-    expect(await caughtError.future, true);
+    expect(uncaught, false);
+  });
+
+  test('ImageProvider.resolve errors in the completer will be caught', () async {
```
Comment (goderbauer) [inline packages/flutter/test/painting/image_provider_test.dart]: Is this an accidental duplication of the previous test or is there something different here?

**Review by goderbauer — APPROVED** [review]
LGTM with one comment

Is the package:test bump related to this? Makes it hard to review :(

**Comment by jonahwilliams** [comment]
Unfortunately I had to run update-packages so I could get the deps for the new integration test :( 

I should probably fix the script so that is not necessary....

File: packages/flutter/test/painting/image_provider_test.dart (hunk 5 above)
Comment (jonahwilliams) [reply packages/flutter/test/painting/image_provider_test.dart]: Woops

**Comment by jonahwilliams** [comment]
The last test did complete successfully,

**Pull request merged**